File: ImageResize/ImageResize/ImageViewController.m
//
//  ImageViewController.m
//  ImageResize
//
//  Created by kuroky on 2019/5/22.
//  Copyright © 2019 Emucoo. All rights reserved.
//

#import "ImageViewController.h"
#import "UIImage+CoreGraphicsAdd.h"
#import "UIImage+CoreImageAdd.h"
#import "UIImage+UIKitAdd.h"
#import "UIImage+ImageIOAdd.h"
#import "UIImage+vImageAdd.h"

static NSInteger const  kTotalCount     =   20;

@interface ImageViewController ()

@property (weak, nonatomic) IBOutlet UIImageView *imageView;
@property (weak, nonatomic) IBOutlet UIButton *clickButton;
@property (nonatomic, strong) NSURL *imgUrl;
@property (nonatomic, assign) CGSize size;
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *activityView;

@end

@implementation ImageViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = [UIColor whiteColor];
    self.navigationItem.title = @"Image";
    
    [self.clickButton setTitle:@"Resize" forState:UIControlStateNormal];
    [self.clickButton setTitle:@"Resizing" forState:UIControlStateSelected];
    
    self.activityView.hidden = YES;
    self.imgUrl = [[NSBundle mainBundle] URLForResource:@"VIIRS_3Feb2012_lrg" withExtension:@"jpg"];
}

- (void)viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];
    
    CGFloat scale = UIScreen.mainScreen.scale;
    CGAffineTransform transform = CGAffineTransformMakeScale(scale, scale);
    self.size = CGSizeApplyAffineTransform(self.imageView.bounds.size, transform);
}

- (IBAction)clickResize:(UIButton *)sender {
    sender.selected = YES;
    sender.userInteractionEnabled = NO;
    self.imageView.image = nil;
    self.activityView.hidden = NO;
    [self.activityView startAnimating];
    
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        CFTimeInterval start = CACurrentMediaTime();
        UIImage *image = [self resizeImage];
        dispatch_async(dispatch_get_main_queue(), ^{
            CGFloat duration = 1.0;
            [UIView transitionWithView:self.imageView duration:duration options:UIViewAnimationCurveEaseOut|UIViewAnimationOptionTransitionCrossDissolve
                            animations:^{
                                self.imageView.image = image;
                            }
                            completion:^(BOOL finished) {
                                [self.activityView stopAnimating];
                                self.activityView.hidden = YES;
                                sender.selected = NO;
                                sender.userInteractionEnabled = YES;
                                CFTimeInterval end = CACurrentMediaTime();
                                NSLog(@"%@ time: %.2fs", [self resizeMethod], end - start - duration);
                            }];
        });
    });
}

- (IBAction)clickTest:(UIButton *)sender {
    sender.userInteractionEnabled = NO;
    [sender setTitle:@"Testing..." forState:UIControlStateNormal];
    
    self.activityView.hidden = NO;
    [self.activityView startAnimating];
    
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        CFTimeInterval start = CACurrentMediaTime();
        for (NSInteger i = 0; i < kTotalCount; i++) {
            [self resizeImage];
        }
        
        CFTimeInterval end = CACurrentMediaTime();
        dispatch_async(dispatch_get_main_queue(), ^{
            sender.userInteractionEnabled = YES;
            [sender setTitle:@"DoTest" forState:UIControlStateNormal];
             NSLog(@"test time: %.2fs", (end - start) / kTotalCount);
            
            self.activityView.hidden = YES;
            [self.activityView stopAnimating];
        });
    });
}

- (UIImage *)resizeImage {
    if (self.resizeType == ResizeWithCoreGraphics) {
        return [UIImage cg_resizedImage:self.imgUrl forSize:self.size];
    }
    else if (self.resizeType == ResizeWithCoreImage) {
        return [UIImage ci_resiImage:self.imgUrl forSize:self.size];
    }
    else if (self.resizeType == ResizeWithUIKit) {
        return [UIImage uikit_resiImage:self.imgUrl forSize:self.size];
    }
    else if (self.resizeType == ResizeWithImageIO) {
        return [UIImage imageio_resiImage:self.imgUrl forSize:self.size];
    }
    else if (self.resizeType == ResizeWithImageIO1) {
        return [UIImage imageio_resiImageWithHintingAndSubsampling:self.imgUrl forSize:self.size];
    }
    else if (self.resizeType == ResizeWithvImage) {
        return [UIImage v_resiImage:self.imgUrl forSize:self.size];
    }
    return nil;
}

- (NSString *)resizeMethod {
    if (self.resizeType == ResizeWithCoreGraphics) {
        return @"Core Graphics";
    }
    else if (self.resizeType == ResizeWithCoreImage) {
        return @"CoreImage";
    }
    else if (self.resizeType == ResizeWithImageIO) {
        return @"ImageIO";
    }
    else if (self.resizeType == ResizeWithImageIO1) {
        return @"ImageIO1";
    }
    else if (self.resizeType == ResizeWithUIKit) {
        return @"UIKit";
    }
    else if (self.resizeType == ResizeWithvImage) {
        return @"vImage";
    }
    return nil;
}

@end

File: Image-Resizing-Example-master/Example/ImageViewController.swift
//
//  ImageViewController.swift
//  Image Resizing
//
//  Created by kuroky on 2019/5/29.
//  Copyright © 2019 NSHipster. All rights reserved.
//

import UIKit

enum ResizeType: Int {
    case ResizeWithCoreGraphics              =           1
    case ResizeWithCoreImage                 =           2
    case ResizeWithImageIO                   =           3
    case ResizeWithImageIO1                  =           6
    case ResizeWithUIKit                     =           4
    case ResizeWithvImage                    =           5
}


class ImageViewController: UIViewController {

    @IBOutlet weak var imageView: UIImageView!
    @IBOutlet weak var activityView: UIActivityIndicatorView!
    @IBOutlet weak var resizeButton: UIButton!
    @IBOutlet weak var testButton: UIButton!
    
    var imgUrl: URL!
    var resizeType: Int?
    var size: CGSize!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.navigationItem.title = "Image"
        
        self.resizeButton.setTitle("Resize", for: .normal)
        self.resizeButton.setTitle("Resizing", for: .selected)
        
        self.activityView.isHidden = true
        self.imgUrl = Bundle.main.url(forResource: "VIIRS_3Feb2012_lrg", withExtension: "jpg")
        
    }
    
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        
        let scale = UIScreen.main.scale;
        let transform = CGAffineTransform.init(scaleX: scale, y: scale)
        self.size = self.imageView.bounds.size.applying(transform)
    }
    
    
    @IBAction func clickResize(_ sender: UIButton) {
        sender.isSelected = true
        sender.isUserInteractionEnabled = false
        self.imageView.image = nil
        self.activityView.isHidden = false
        self.activityView.startAnimating()
        
        DispatchQueue.global().async {
            let start = CACurrentMediaTime()
            let image = self.resizeImage()
            
            DispatchQueue.main.async {
                let duration = 1.0;
                UIView.transition(with: self.imageView,
                                  duration: duration,
                                  options: .curveEaseOut,
                                  animations: {
                                    self.imageView.image = image
                }, completion: { _ in
                    self.activityView.stopAnimating()
                    self.activityView.isHidden = true
                    sender.isSelected = false
                    sender.isUserInteractionEnabled = true
                    let end = CACurrentMediaTime()
                    print("%@ time: %.2fs", self.resizeMethod() as String, end - start - duration);
                })
            }
        }
    }
    
    @IBAction func clickTest(_ sender: UIButton) {
    
    }
        
    func resizeImage() -> UIImage? {
        if (self.resizeType == ResizeType.ResizeWithCoreGraphics.rawValue) {
            return CoreGraphics.resizedImage(at: self.imgUrl, for: self.size)
        }
        else if (self.resizeType == ResizeType.ResizeWithCoreImage.rawValue) {
            return CoreImage.resizedImage(at: self.imgUrl, for: self.size)
        }
        else if (self.resizeType == ResizeType.ResizeWithUIKit.rawValue) {
            return UIKit.resizedImage(at: self.imgUrl, for: self.size)
        }
        else if (self.resizeType == ResizeType.ResizeWithImageIO.rawValue) {
            return ImageIO.resizedImage(at: self.imgUrl, for: self.size)
        }
        else if (self.resizeType == ResizeType.ResizeWithImageIO1.rawValue) {
            return ImageIO.resizedImageWithHintingAndSubsampling(at: self.imgUrl, for: self.size)
        }
        else if (self.resizeType == ResizeType.ResizeWithvImage.rawValue) {
            return vImage.resizedImage(at: self.imgUrl, for: self.size)
        }
        return nil;
    }

    func resizeMethod() -> String! {
        if (self.resizeType == ResizeType.ResizeWithCoreGraphics.rawValue) {
            return "Core Graphics";
        }
        else if (self.resizeType == ResizeType.ResizeWithCoreImage.rawValue) {
            return "CoreImage";
        }
        else if (self.resizeType == ResizeType.ResizeWithImageIO.rawValue) {
            return "ImageIO";
        }
        else if (self.resizeType == ResizeType.ResizeWithImageIO1.rawValue) {
            return "ImageIO1";
        }
        else if (self.resizeType == ResizeType.ResizeWithUIKit.rawValue) {
            return "UIKit";
        }
        else if (self.resizeType == ResizeType.ResizeWithvImage.rawValue) {
            return "vImage";
        }
        return nil;
    }
}
